keps/0019-package-api-versioning.md

```diff
@@ -2,64 +2,374 @@
-### Support multiple API versions in the package repository
+## Proposals
+
+### With regards to package version and operator version
+
+Two proposals: singular and composite package/operator version. In the case of potential operator "bundles" (operators containing multiple app versions) both would end up "hiding" those app versions behind the package/operator version.
+
+#### Singular package/operator version (largely, the current state of things)
+
+##### Pros
+
+- Visually simple
+- It can be known just from the package version whether or not it contains a certain operator-related feature or bug fix
+
+##### Cons
+
+- Operator users care about the application version, not a "synthetic" version
+- Makes it complex to have an "app version"-based backport strategy. Customers might not be able to use operator based on newer app versions due to constraints outside of our control, but would still want to receive new operator-related features and bug fixes
+- Continuing from the second "pro", it might be possible that subsequent package versions with different app versions don't contain an operator-related feature or bug fix that only makes sense for a specific app version (e.g., version N+1 based on app version M doesn't include an operator feature available on version N based on app version M+1 because the operator feature only makes sense for app version M)
+- The progression of the package version is confusing with regards to the application version progression
+
+  Example:
+
+  | Time | Package/operator version | App version        | Comment                                            |
+  | ---- | ---------------          | ------------------ | -------------------------------------------------- |
+  | T0   | 1.0.0                    | 2.3.0              | Initial release based on app version 2.3.x         |
+  | T1   | 1.1.0                    | 3.0.0              | Release based on app version 3.0.x                 |
+  | T2   | 1.2.0                    | 2.3.0              | New operator-related feature for app version 2.3.x |
+  | T2   | 1.3.0                    | 3.0.0              | New operator-related feature for app version 3.0.x |
+  | T3   | 1.4.0                    | 2.3.1              | Bug fix release for app version 2.3.x              |
+  | T4   | 1.5.0                    | 2.3.1              | Important bug fix A in operator-related code       |
+  | T4   | 1.6.0                    | 3.0.0              | Important bug fix A in operator-related code       |
+  | T5   | 1.6.1                    | 2.3.1              | Small bug fix B in operator-related code           |
+  | T5   | 1.6.2                    | 3.0.0              | Small bug fix B in operator-related code           |
+
+  The example above illustrates two sources of confusion:
+  - A higher package version can contain a lower app version
+  - There's no relation between the package version growth and the app version growth
+
+#### Composite package/operator version
```

Comment:
can you elaborate somewhere how will this look like on a package with *no* underlying application?

Reply:
That's a great point! I didn't have time to elaborate on it yesterday, but added an entry to my todo list for it. Will do it today.